spec/message_spec.rb

```diff
@@ -0,0 +1,67 @@
+RSpec.describe "message encoding" do
+  # message Sally {
+  #   required string mood = 1;
+  #   optional uint32 age  = 2;
+  # }
+  class Sally
+    def initialize(attributes)
+      @attributes = {}
+      attributes.each do |attribute, value|
+        public_send("#{attribute}=", value)
+      end
+    end
+
+    ## Attributes
+
+    def age
+      @attributes[:age]
+    end
+
+    def age=(age)
+      raise ArgumentError, "age is not a number" unless age.respond_to?(:to_i)
+      raise ArgumentError, "age is too large" unless age.to_i <= 4_294_967_295
```

Comment:
+1 for delaying attribute cohersion until encoding.  I think that cohersion in the setter leads to some unexpected surprises if you try to do anything to your attributes after you have set them.

```
f = Foo.new # A protobuf message object
f.awesome_int_field = 6.0

# looks like this should be 6.0, but if you coherce in the setter this would be 6!
f.awesome_int_field * 1

# This one is a bit out of left field, but starts to get weird 
f.awesome_int_field = "1"
f.awesome_int_field.strip!
```